drivers/flash/flash_andes_qspi.c

```diff
@@ -0,0 +1,948 @@
+#include <zephyr/logging/log.h>
+
+#include "flash_andes_qspi.h"
+#include "spi_nor.h"
+#include "jesd216.h"
+#include "flash_priv.h"
+
+LOG_MODULE_REGISTER(flash_andes, CONFIG_FLASH_LOG_LEVEL);
+
+/* Indicates that an access command includes bytes for the address.
+ * If not provided the opcode is not followed by address bytes.
+ */
+#define ANDES_ACCESS_ADDRESSED BIT(0)
+
+/* Indicates that an access command is performing a write.  If not
+ * provided access is a read.
+ */
+#define ANDES_ACCESS_WRITE BIT(7)
+
+typedef void (*flash_andes_qspi_config_func_t)(void);
+
+struct flash_andes_qspi_config {
+	flash_andes_qspi_config_func_t cfg_func;
+	uint32_t base;
+	uint32_t irq_num;
+	struct flash_parameters parameters;
+	bool xip;
+#if defined(CONFIG_FLASH_ANDES_QSPI_SFDP_DEVICETREE)
+	uint8_t jedec_id[SPI_NOR_MAX_ID_LEN];
+	uint32_t flash_size;
+	uint8_t bfp_len;
+	const struct jesd216_bfp *bfp;
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+	struct flash_pages_layout layout;
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+#endif /* CONFIG_FLASH_ANDES_QSPI_SFDP_DEVICETREE */
+};
+
+struct flash_andes_qspi_data {
+	struct k_sem sem;
+	struct k_sem device_sync_sem;
+	uint32_t tx_fifo_size;
+	uint32_t rx_fifo_size;
+	uint8_t	 *tx_buf;
+	uint8_t	 *rx_buf;
+	uint32_t tx_len;
+	uint32_t rx_len;
+	uint32_t tx_ptr; /* write pointer  */
+	uint32_t rx_ptr; /* read pointer  */
+	struct jesd216_erase_type erase_types[JESD216_NUM_ERASE_TYPES];
+	uint16_t page_size;
+#ifdef CONFIG_FLASH_ANDES_QSPI_SFDP_RUNTIME
+	uint32_t flash_size;
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+	struct flash_pages_layout layout;
+#endif /* CONFIG_FLASH_PAGE_LAYOUT */
+#endif
+};
+
+static int flash_andes_qspi_write_protection_set(const struct device *dev,
+					bool write_protect);
+
+/* Get pointer to array of supported erase types. */
+static inline const struct jesd216_erase_type *
+dev_erase_types(const struct device *dev)
+{
+	const struct flash_andes_qspi_data *data = dev->data;
+
+	return data->erase_types;
+}
+
+/* Get the size of the flash device. */
+static inline uint32_t dev_flash_size(const struct device *dev)
+{
+#ifdef CONFIG_FLASH_ANDES_QSPI_SFDP_RUNTIME
+	const struct flash_andes_qspi_data *data = dev->data;
+
+	return data->flash_size;
+#else /* CONFIG_FLASH_ANDES_QSPI_SFDP_RUNTIME */
+	const struct flash_andes_qspi_config *cfg = dev->config;
+
+	return cfg->flash_size;
+#endif /* CONFIG_FLASH_ANDES_QSPI_SFDP_RUNTIME */
+}
+
+/* Get the flash device page size. */
+static inline uint16_t dev_page_size(const struct device *dev)
+{
+	const struct flash_andes_qspi_data *data = dev->data;
+
+	return data->page_size;
+}
+
+/*
+ * @brief Send an SPI command
+ *
+ * @param dev Device struct
+ * @param opcode The command to send
+ * @param access flags that determine how the command is constructed.
+ * @param addr The address to send
+ * @param data The buffer to store or read the value
+ * @param length The size of the buffer
+ * @return 0 on success, negative errno code otherwise
```

Comment:
As I haven't identified the specific error case that this function should handle, I have temporarily set it to always return 0.
I've temporarily removed the comment ', negative errno code otherwise' until I identify the case that requires the function to return an error code.